File: erlang/strain/src/strain.erl
-module(strain).

-export([keep/2, discard/2, test_version/0]).

keep(Fn, List) ->
  keep(Fn, List, []).

keep(Fn, [], Result) ->
  lists:reverse(Result);

keep(Fn, [Head | Tail], Result) ->
  case Fn(Head) of
    true  -> keep(Fn, Tail, [Head | Result]) ;
    false -> keep(Fn, Tail, Result)
  end.

discard(Fn, List) ->
  keep(fun (X) -> not Fn(X) end, List).

test_version() ->
  1.

File: elixir/strain/strain.exs
defmodule Strain do
  @spec keep(list :: list(any), fun :: ((any) -> boolean)) :: list(any)
  def keep(list, fun),
    do: do_keep(list, fun, [])

  @spec discard(list :: list(any), fun :: ((any) -> boolean)) :: list(any)
  def discard(list, fun) do
    keep(list, fn x -> !fun.(x) end)
  end

  defp do_keep([], _fun, result),
    do: Enum.reverse(result)

  defp do_keep([h | t], fun, result) do
    case fun.(h) do
      true  -> do_keep(t, fun, [h | result])
      false -> do_keep(t, fun, result)
    end
  end
end
